File: linked-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int count = 0;
        ListNode curr = head;
        while (curr != null) {
            count++;
            curr = curr.next;
        }
        int pos = count - n + 1;
        if (pos == 1) {
            return head.next;
        }
        count = 1;
        curr = head;
        while (curr != null) {
            if (count == pos - 1) {
                curr.next = curr.next.next;
            }
            curr = curr.next;
            count++;
        }
        return head;
    }
}

File: linked-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        cur = head
        count = 0
        while cur:
            cur = cur.next
            count += 1
        if count == n:
            return None
        curr = head
        pos = 0
        while curr:
            pos += 1
            if pos == count - n:
                curr.next = curr.next.next

            curr = curr.next
        return head
